flow/aggregate.py

```diff
@@ -0,0 +1,183 @@
+# Copyright (c) 2020 The Regents of the University of Michigan
+# All rights reserved.
+# This software is licensed under the BSD 3-Clause License.
+from collections.abc import Iterable
+from copy import deepcopy
+from itertools import groupby
+from itertools import zip_longest
+from functools import partial
+
+import signac
+from signac.contrib.project import Project, JobsCursor
+
+
+class aggregate:
+    """Decorator for operation functions that are to be aggregated.
+
+    If this class is used for aggregation then by-default, if the aggregator
+    parameter is not passed, an aggregate of all jobs will be created.
+
+    .. code-block:: python
+
+        example_aggregate = aggregate()
+        @example_aggregate
+        @FlowProject.operation
+        def foo(*jobs):
+            print(len(jobs))
+
+    :param aggregator:
+        Information on how to aggregate jobs. Takes in a list of
+        jobs and can return or yield lists or single jobs instead.
+        The default behavior is creating an aggregate of all jobs.
+    :type aggregator:
+        callable
+    :param sort:
+        Before aggregating, sort the jobs given by a statepoint parameter.
+        The default value is None.
+    :type sort:
+        str or NoneType
+    :param reverse:
+        States if the jobs are to be sorted in reverse order.
+        The default value is False.
+    :type reverse:
+        bool
+    :param select:
+        Condition for filtering jobs. This operates on a single job.
+        The default value is None
+    :type select:
+        callable or NoneType
+    """
+
+    def __init__(self, aggregator=None, sort=None, reverse=False, select=None):
+        if aggregator is None:
+            def aggregator(jobs):
+                return jobs
+
+        if not callable(aggregator):
+            raise TypeError("Expected callable aggregator function, got {}"
+                            "".format(type(aggregator)))
+
+        if sort is not None and not isinstance(sort, str):
+            raise TypeError("Expected string sort parameter, got {}".format(type(sort)))
+
+        if not isinstance(reverse, bool):
+            raise TypeError("Expected bool reverse parameter got {}".format(type(reverse)))
```

Comment:
Here I would be lenient we could just do `reverse = bool(reverse)`, so 0 or 1 would work.